### PR TITLE
Ensure borrows of fn/closure params do not outlive invocations

### LINKED ISSUE #29793: rustc fails to produce lifetime and closure related compiler error.
We ran into this in conrod, where one of the widgets would produce different behaviour depending on whether it was compiled in debug or release.

The related conrod issue is [here](https://github.com/PistonDevelopers/conrod/issues/619).

[This](https://github.com/PistonDevelopers/conrod/issues/619#issuecomment-155974672) comment describes the strange behaviour more precisely.

[This](https://github.com/PistonDevelopers/conrod/issues/619#issuecomment-155977384) comment demonstrates the fix used.

@huonw and @Aatch helped to point out that the difference in behaviour seemed to be due to the fact that rustc was failing to produce an error relating to the lifetimes of the `demo`, `col` and `row` values.

@Aatch shared a simplified example of what error we *should* have received [here](https://play.rust-lang.org/?gist=c490d7550c8cfc310612&version=stable).

Edit: I should mention, this bug was found in:
- `rustc 1.4.0 (8ab8581f6 2015-10-27)`
- `rustc 1.6.0-nightly (5b4986fa5 2015-11-08)`
- `rustc 1.5.0-beta.2 (b0f440163 2015-10-28)`

on both windows and osx.

### PR DESCRIPTION
**Pull request by pnkfelix**
Ensure borrows of fn/closure params do not outlive invocations.

Does this by adding a new CallSiteScope to the region (or rather code extent) hierarchy, which outlives even the ParameterScope (which in turn outlives the DestructionScope of a fn/closure's body).

Fix #29793 

r? @nikomatsakis

### CODE REVIEW

**Comment by pnkfelix** [comment]
@nikomatsakis by the way, as previously discussed, this PR removed `BlockScope` entirely (and puts in `FnScope` instead) ... but after doing so, I did think of one potential reason why we might keep `BlockScope`, or put it back in in the future.  Namely, if we eventually adopt the ability to name the lifetimes of blocks, e.g. `'a: { ... 'b: { ... } ... }`, then we will probably need `BlockScope` again (i think...).  Or maybe I'm missing something.

Anyway, we don't seem to need it *right now*, so I took it out.

File: src/librustc/middle/resolve_lifetime.rs [hunk 1]
```diff
@@ -488,7 +481,10 @@ impl<'a> LifetimeContext<'a> {
         // `self.labels_in_fn`.
         extract_labels(self, fb);
 
-        self.visit_block(fb);
+        self.with(FnScope { fn_id: fn_id, body_id: fb.id, s: self.scope },
+                  move |_old_scope, this| {
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: do you need `move` here? I wouldn't expect so...?

File: src/librustc/middle/resolve_lifetime.rs [hunk 2]
```diff
@@ -559,8 +555,10 @@ impl<'a> LifetimeContext<'a> {
         let mut scope = self.scope;
         loop {
             match *scope {
-                BlockScope(blk_scope, s) => {
-                    return self.resolve_free_lifetime_ref(blk_scope, lifetime_ref, s);
+                FnScope {fn_id, body_id, s } => {
+                    return self.resolve_free_lifetime_ref(
+                        region::CallSiteScopeData { fn_id: fn_id, body_id: body_id },
+                        lifetime_ref, s);
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: Nit: newline before `s`

File: src/librustc/middle/ty/mod.rs [hunk 3]
```diff
@@ -1196,11 +1197,18 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
                             MethodTraitItem(ref method_ty) => {
                                 let method_generics = &method_ty.generics;
                                 let method_bounds = &method_ty.predicates;
+                                let extent = if id != body_id {
```
Comment (nikomatsakis) [inline src/librustc/middle/ty/mod.rs]: I suspect `body_id` is not used elsewhere... probably cleaner to write `body.is_none()` or `match body` here.

File: src/librustc/middle/ty/mod.rs [hunk 4]
```diff
@@ -2623,20 +2629,22 @@ impl<'tcx> ctxt<'tcx> {
...
                                                -> ParameterEnvironment<'a, 'tcx>
     {
         //
         // Construct the free substs.
         //
 
-        let free_substs = self.construct_free_substs(generics, free_id);
-        let free_id_outlive = self.region_maps.item_extent(free_id);
+        let free_substs = self.construct_free_substs(generics, free_id_outlive);
+        // let free_id_outlive = self.region_maps.item_extent(free_id);
```
Comment (nikomatsakis) [inline src/librustc/middle/ty/mod.rs]: Nit: remove commented line

File: src/librustc_typeck/check/regionck.rs [hunk 5]
```diff
@@ -315,9 +327,19 @@ impl<'a, 'tcx> Rcx<'a, 'tcx> {
...
         self.visit_region_obligations(body.id);
 
+        let call_site_scope = self.call_site_scope.unwrap();
+        debug!("visit_fn_body body.id {} call_site_scope: {:?}",
+               body.id, call_site_scope);
+        type_of_node_must_outlive(self,
+                                  infer::CallReturn(span),
+                                  body.id,
+                                  ty::ReScope(call_site_scope)
+                                  );
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/regionck.rs]: Nit: unusual formatting of `)`

File: src/librustc_typeck/check/regionck.rs [hunk 6]
```diff
@@ -834,6 +856,18 @@ fn visit_expr(rcx: &mut Rcx, expr: &hir::Expr) {
...
 
+        hir::ExprRet(Some(ref ret_expr)) => {
+            let call_site_scope = rcx.call_site_scope;
+            debug!("visit_expr ExprRet ret_expr.id {} call_site_scope: {:?}",
+                   ret_expr.id, call_site_scope);
+            type_of_node_must_outlive(rcx,
+                                      infer::CallReturn(ret_expr.span),
+                                      ret_expr.id,
+                                      ty::ReScope(call_site_scope.unwrap())
+                                      );
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/regionck.rs]: Nit: unusual formatting of `)`

**Comment by nikomatsakis** [comment]
r+ modulo nits.

**Comment by nikomatsakis** [comment]
Marking as relnotes since this is a soundness fix that did cause some amount of downstream fallout (e.g., #30519).

**Pull request merged**